### PR TITLE
feat(cache): remove persistent raw values from queue

### PR DESCRIPTION
**Pull request by uael**
This effort aims to optimize database load by reducing duplication of persistent raw values (e.g., code content, lock file, or flow value) and leveraging caching wherever possible.

## Implementation

- **Flow Job**:  
  - The `raw_flow` is no longer directly inserted. Instead, the latest version ID of the flow is stored.  
  - On pull, the flow value is fetched using the version ID (which is unique across entities) with caching. The "lite" version is fetched first, falling back to the full version if not present (to prevent a race condition when deploying a new flow).

- **Flow Dependencies Job**:  
  - The `raw_flow` is no longer directly inserted. Instead, the version ID already stored is used on pull to retrieve the flow value with caching.  
  - Race conditions during dependency handling were resolved. Insert/update operations for new flow values are now performed within the same transaction as lock generation.

- **Cache**:  
  - All raw values are now retrieved via a new file-system-backed cache, significantly reducing raw value re-allocations (e.g., `.clone()`).

## Backward Compatibility

To support older workers (pre-`v1.440`):  
- If `MIN_VERSION_IS_AT_LEAST_1_440 == false`, the previous behavior is retained (raw values are inserted on push).  
- Starting from `v1.440`, once all workers are updated, raw values will no longer be inserted during job pushes.  
- **Caveat**: If an outdated worker starts after assuming all workers are up to date (`MIN_VERSION_IS_AT_LEAST_1_440 == true`), it may fail to process jobs correctly during the brief period before `MIN_VERSION_IS_AT_LEAST_1_440` is updated.

## Tests

Manual testing has been performed extensively. Additionally, new integration tests have been added to prevent regressions.

> [!IMPORTANT]
> Introduces caching for raw values, removing them from database tables and storing them in a cache system, with updates to job, flow, and script handling.
> 
>   - **Behavior**:
>     - Removes `raw_code`, `raw_lock`, `raw_flow` from `queue` and `job` tables, storing them in a cache system.
>     - Updates job handling to fetch raw values from cache instead of database.
>   - **Caching**:
>     - Implements caching mechanisms in `cache.rs` for flows, scripts, and app data.
>     - Introduces `FlowData` and `ScriptData` structs for managing cached raw values.
>     - Adds `fetch_flow()`, `fetch_version()`, `fetch_version_lite()`, and `fetch_script()` functions for caching.
>   - **Flow and Script Handling**:
>     - Updates flow and script dependency handling to use cached values.
>     - Improves efficiency of raw value management.
>   - **Misc**:
>     - Adds new SQL queries for fetching cached data.
>     - Improves error handling for cache misses.
>     - Updates worker code to handle cached values.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=windmill-labs%2Fwindmill&utm_source=github&utm_medium=referral)<sup> for ec32db6ae0ee443b5d8de973faab813133409005. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: backend/windmill-queue/src/jobs.rs [hunk 1]
```diff
@@ -3572,13 +3584,22 @@ pub async fn push<'c, 'd>(
                     }))
                 });
             }
+
             let status = Some(FlowStatus::new(&value));
+            // Keep inserting `value` if not all workers are updated.
+            // Starting at `v1.436`, the value is fetched on pull from the version id.
+            let value_o = if !*MIN_VERSION_IS_AT_LEAST_1_436.read().await {
+                Some(value.clone())
```
Comment (uael) [inline backend/windmill-queue/src/jobs.rs]: No clone

File: backend/windmill-common/src/cache.rs [hunk 2]
```diff
@@ -21,24 +29,31 @@ pub struct FsBackedCache<Key, Val> {
...
+        F: Future<Output = error::Result<T>>,
     {
         self.cache
-            .get_or_insert_async(
-                key,
-                fs::import_or_insert_with(self.root, (*key).into(), with),
-            )
+            .get_or_insert_async(key, async {
+                let key = key.to_owned();
+                fs::import_or_insert_with(&key.path(&Path::new(self.root)), with)
```
Comment (uael) [inline backend/windmill-common/src/cache.rs]: Just self.root

File: backend/windmill-common/src/auth.rs [hunk 3]
```diff
@@ -75,7 +75,7 @@ pub async fn is_super_admin_email(db: &DB, email: &str) -> Result<bool> {
 
 pub async fn is_devops_email(db: &DB, email: &str) -> Result<bool> {
     if is_super_admin_email(db, email).await? {
-        return Ok(true)
+        return Ok(true);
```
Comment (uael) [inline backend/windmill-common/src/auth.rs]: Re-add

File: backend/windmill-common/src/cache.rs [hunk 4]
```diff
@@ -161,144 +295,121 @@ pub mod flow {
...
+            .get_or_insert_async(
+                &node,
+                |(raw_script, raw_flow)| match raw_flow {
+                    Some(raw_flow) => RawData::Flow(Arc::new(raw_flow)),
+                    None => RawData::Script(Arc::new(raw_script)),
+                },
+                async {
+                    sqlx::query!(
+                        "SELECT \
                         lock AS \"lock: String\", \
```
Comment (uael) [inline backend/windmill-common/src/cache.rs]: Indent

File: backend/windmill-common/src/cache.rs [hunk 5]
```diff
@@ -418,69 +472,136 @@ pub mod app {
...
         id: AppScriptId,
-    ) -> error::Result<(Option<String>, String)> {
+    ) -> error::Result<Arc<RawScript>> {
         // If not present, `get_or_insert_async` will lock the key until the future completes,
         // so only one thread will be able to fetch the data from the database and write it to
         // the file system and cache, hence no race on the file system.
         CACHE
-            .get_or_insert_async(&id, async {
+            .get_or_insert_async(&id, |x| Arc::new(x), async {
                 sqlx::query!(
```
Comment (uael) [inline backend/windmill-common/src/cache.rs:492]: Just Arc::new

File: backend/windmill-common/src/cache.rs [hunk 6]
```diff
@@ -418,69 +472,136 @@ pub mod app {
...
+                .map(RawData::Script),
+            (Script | Dependencies, Some(hash)) => script::fetch(e, ScriptHash(hash))
+                .await
+                .map(|(raw_script, _metadata)| RawData::Script(raw_script)),
+            (AppScript, Some(id)) => app::fetch_script(e, AppScriptId(id))
+                .await
+                .map(RawData::Script),
+            _ => {
+                return Err(error::Error::InternalErr(format!(
+                    "Invalid job payload job kind: {:?}",
```
Comment (uael) [inline backend/windmill-common/src/cache.rs]: payload for* job kind

File: backend/windmill-worker/src/worker_flow.rs [hunk 7]
```diff
@@ -1495,12 +1495,35 @@ pub async fn handle_flow(
...
 ) -> anyhow::Result<()> {
-    let flow = flow_value
+    let mut flow = flow_value
         .with_context(|| "Unable to parse flow definition")?;
     let status = flow_job
         .parse_flow_status()
         .with_context(|| "Unable to parse flow status")?;
 
+    // Update value.
+    if matches!(flow_job.job_kind, JobKind::Flow) {
```
Comment (uael) [inline backend/windmill-worker/src/worker_flow.rs]: Check if this mutation is really necessary; same_worker can be taken from job and skipping the preprocessor can be done later on based on this same next condition

File: backend/windmill-queue/src/jobs.rs [hunk 8]
```diff
@@ -3512,47 +3517,35 @@ pub async fn push<'c, 'd>(
...
-                    tx
-                )
-            }?
-            .ok_or_else(|| Error::InternalErr(format!("not found flow at path {:?}", path)))?;
-            let mut value = serde_json::from_str::<FlowValue>(value_json.get()).map_err(|err| {
-                Error::InternalErr(format!(
-                    "could not convert json to flow for {path}: {err:?}"
-                ))
-            })?;
+                cache::flow::fetch_version_lite(_db, version).await
```
Comment (uael) [inline backend/windmill-queue/src/jobs.rs]: add fallback to no lite

File: backend/windmill-worker/src/common.rs [hunk 9]
```diff
@@ -640,9 +640,7 @@ pub async fn cached_result_path(
     db: &DB,
     client: &AuthedClient,
     job: &QueuedJob,
-    raw_code: Option<&String>,
-    raw_lock: Option<&String>,
-    raw_flow: Option<&Json<Box<RawValue>>>,
+    raw_data: Option<&RawData>,
 ) -> String {
```
Comment (uael) [inline backend/windmill-worker/src/common.rs:667]: Remove the else if let None

File: backend/windmill-worker/src/worker.rs [hunk 10]
```diff
@@ -2232,35 +2232,18 @@ pub async fn get_hub_script_content_and_requirements(
...
-}
-
 pub async fn get_script_content_by_hash(
     script_hash: &ScriptHash,
-    w_id: &str,
+    _w_id: &str,
     db: &DB,
 ) -> error::Result<ContentReqLangEnvs> {
-    let script = cache::script::fetch(db, *script_hash, w_id).await?;
+    let (raw_script, metadata) = cache::script::fetch(db, *script_hash).await?;
```
Comment (uael) [inline backend/windmill-worker/src/worker.rs]: raw_script -> data

File: backend/windmill-worker/src/worker.rs [hunk 11]
```diff
@@ -2286,80 +2268,90 @@ async fn handle_code_execution_job(
...
-                envs: None,
-                codebase: None,
-            }
+            arc_data = cache::app::fetch_script(db, AppScriptId(script_hash()?.0)).await?;
+            metadata = ScriptMetadata { language: job.language, envs: None, codebase: None };
+            (arc_data.as_ref(), &metadata)
         }
         JobKind::DeploymentCallback => {
-            get_script_content_by_path(job.script_path.clone(), &job.workspace_id, db).await?
+            // FIXME(uael): If found no usage of `JobKind::DeploymentCallback` as a job,
```
Comment (uael) [inline backend/windmill-worker/src/worker.rs]: I* found

File: backend/windmill-queue/src/jobs.rs [hunk 12]
```diff
@@ -3223,12 +3235,29 @@ pub async fn push<'c, 'd>(
...
+                    value.same_worker = true;
+                }
+                if !apply_preprocessor {
+                    value.preprocessor_module = None;
+                }
+                Some(value)
+            } else {
+                // `raw_flow` is fetched on pull, the mutations from the other branch are replaced
+                // by additional checks when handling the flow.
+                None
```
Comment (hugocasa) [inline backend/windmill-queue/src/jobs.rs:3258]: if !apply_preprocessor, you should remove preprocessor from the flow status

File: backend/windmill-queue/src/jobs.rs [hunk 13]
```diff
@@ -3280,18 +3309,22 @@ pub async fn push<'c, 'd>(
...
+            let concurrent_limit = raw_flow.concurrent_limit;
+            let concurrency_time_window_s = raw_flow.concurrency_time_window_s;
+            let cache_ttl = raw_flow.cache_ttl.map(|x| x as i32);
             (
                 None,
                 flow_path,
                 None,
-                JobKind::Flow,
-                Some(raw_flow.clone()),
+                JobKind::FlowPreview,
```
Comment (hugocasa) [inline backend/windmill-queue/src/jobs.rs]: should use JobKind::Flow

**Review by HugoCasa — CHANGES_REQUESTED** [review]

File: backend/windmill-queue/src/jobs.rs [hunk 14]
```diff
@@ -3152,13 +3169,15 @@ pub async fn push<'c, 'd>(
...
+            // this is a new flow being pushed, flow_status is set to flow_value:
+            let flow_status: FlowStatus = FlowStatus::new(&flow_value);
             (
                 None,
                 Some(path),
                 None,
-                JobKind::Flow,
-                Some(flow_value.clone()),
-                Some(FlowStatus::new(&flow_value)), // this is a new flow being pushed, flow_status is set to flow_value
+                JobKind::FlowPreview,
```
Comment (uael) [inline backend/windmill-queue/src/jobs.rs]: Flow

File: backend/windmill-queue/src/jobs.rs [hunk 15]
```diff
@@ -2861,14 +2871,14 @@ pub async fn push<'c, 'd>(
...
-            let status = Some(FlowStatus::new(&value));
+            let data = cache::flow::fetch_flow(_db, id).await?;
+            let status = Some(FlowStatus::new(data.value()?));
             (
                 Some(id.0),
                 Some(path),
                 None,
                 JobKind::FlowNode,
-                Some(value),
+                None, // `raw_flow` is fetched on pull.
```
Comment (uael) [inline backend/windmill-queue/src/jobs.rs]: Check version as well

File: backend/windmill-worker/src/csharp_executor.rs [hunk 16]
```diff
@@ -396,10 +396,7 @@ pub async fn handle_csharp_job(
     let hash = calculate_hash(&format!(
         "{}{}",
         inner_content,
-        requirements_o
-            .as_ref()
-            .map(|x| x.to_string())
-            .unwrap_or_default()
+        requirements_o.map(|x| x.as_str()).unwrap_or("")
```
[the remark this answers is not in the document]
Comment (uael) [reply backend/windmill-worker/src/csharp_executor.rs]: Good bot

File: backend/windmill-worker/src/csharp_executor.rs [hunk 17]
```diff
@@ -396,10 +396,7 @@ pub async fn handle_csharp_job(
     let hash = calculate_hash(&format!(
         "{}{}",
         inner_content,
-        requirements_o
-            .as_ref()
-            .map(|x| x.to_string())
-            .unwrap_or_default()
+        requirements_o.unwrap_or(&String::new())
```
[the remark this answers is not in the document]
Comment (uael) [reply backend/windmill-worker/src/csharp_executor.rs:403]: Nope

**Comment by uael** [comment]
In handle_flow_dependencies, move `tx.commit().await?;` before the lite version insertion

**Pull request merged**